### PR TITLE
Add two new build targets to enable the possibility of using clang-cl as an assembler for Windows on Arm

### PR DESCRIPTION
**Pull request by everton1984**
Currently it is not possible to use the assembler files for Windows on Arm as there is no NASM for aarch64. This patch makes the appropriate changes to use `clang-cl` as an assembler along with the possibility of also using it as a compiler. Forcibly adding `__ASSEMBLER__` was required as cl does not add it by itself.

### CODE REVIEW

File: Configurations/50-win-onecore.conf [hunk 1]
```diff
@@ -82,7 +82,35 @@ my %targets = (
         ex_libs         => "onecore.lib",
         multilib        => "-arm64",
     },
-
+    "VC-WIN64-CLANGASM-ARM" => {
+        inherit_from    => [ "VC-noCE-common" ],
+        defines         => add("_ARM_WINAPI_PARTITION_DESKTOP_SDK_AVAILABLE",
+                               "OPENSSL_SYS_WIN_CORE"),
```
Comment (levitte) [inline Configurations/50-win-onecore.conf:88]: Please add `"__ASSEMBLER__"` here

File: Configurations/50-win-onecore.conf [hunk 2]
```diff
@@ -82,7 +82,35 @@ my %targets = (
...
+        asflags   => "/c",
+        asoutflag => "/Fo",
+        perlasm_scheme => "win64",
+        uplink_arch      => 'armv8',
+    },
+    "VC-CLANG-WIN64-CLANGASM-ARM" => {
+        CC => "clang-cl",
+        inherit_from    => [ "VC-noCE-common" ],
+        defines         => add("_ARM_WINAPI_PARTITION_DESKTOP_SDK_AVAILABLE",
+                               "OPENSSL_SYS_WIN_CORE"),
```
Comment (levitte) [inline Configurations/50-win-onecore.conf:103]: Please add `"__ASSEMBLER__"` here

File: Configurations/windows-makefile.tmpl [hunk 3]
```diff
@@ -773,13 +773,24 @@ EOF
...
         del /Q \$@.S
 EOF
+                } else {
+                    return <<"EOF";
+$target: "$gen0" $deps
+	cmd /C "set "ASM=\$(AS)" & $generator \$@.S"
+    \$(CPP) $incs $cppflags $defs \$@.S > \$@.i
+	move /Y \$@.i \$@
+        del /Q \$@.S
+EOF
```
Comment (levitte) [inline Configurations/windows-makefile.tmpl:792]: Please remove this hackery, the addtions in the target configs will cover them

File: Configurations/windows-makefile.tmpl [hunk 4]
```diff
@@ -773,13 +773,24 @@ EOF
...
               # end up generating foo.s in two steps.
               if ($args{src} =~ /\.S$/) {
-                   return <<"EOF";
+                if(($config{CC} !~ /clang-cl/) and ($config{AS} =~ /clang-cl/))
+                {
+                    return <<"EOF";
 $target: "$gen0" $deps
 	cmd /C "set "ASM=\$(AS)" & $generator \$@.S"
-	\$(CPP) $incs $cppflags $defs \$@.S > \$@.i
+    \$(CPP) /D__ASSEMBLER__ /E $incs $cppflags $defs \$@.S > \$@.i
```
Comment (levitte) [inline Configurations/windows-makefile.tmpl:781]: CPP should already be defined to `$(CC) /EP /C`, why did you need to add `/E` explicitly here?

File: Configurations/50-win-onecore.conf [hunk 5]
```diff
@@ -82,7 +82,35 @@ my %targets = (
         ex_libs         => "onecore.lib",
         multilib        => "-arm64",
     },
-
+    "VC-WIN64-CLANGASM-ARM" => {
```
Comment (levitte) [inline Configurations/50-win-onecore.conf:85]: I would rather have your definitions in a separate file, say `Configurations/50-win-clang-cl.conf`

**Review by levitte — CHANGES_REQUESTED** [review]
Overall a good effort, thank you.  I do have a few smaller nits, though

File: Configurations/50-win-onecore.conf (hunk 1 above)
Comment (everton1984) [reply Configurations/50-win-onecore.conf:88]: @levitte Thanks a lot for the feedback. Adding it directly here will define it for all makefile targets. If I understand the use of the flag correctly it is just enabled to tell the compiler we're preprocessing an assembler file https://gcc.gnu.org/onlinedocs/gcc-5.1.0/cpp/Standard-Predefined-Macros.html which is not always true. This is the reason for the hack on the template file.

File: Configurations/windows-makefile.tmpl [hunk 6]
```diff
@@ -773,13 +773,24 @@ EOF
               # If the target is named foo.S in build.info, we want to
               # end up generating foo.s in two steps.
               if ($args{src} =~ /\.S$/) {
-                   return <<"EOF";
+                if(($config{CC} !~ /clang-cl/) and ($config{AS} =~ /clang-cl/))
+                {
+                    return <<"EOF";
```
Comment (levitte) [inline Configurations/windows-makefile.tmpl:778]: This is unnecessary, as the added `/D__ASSEMBLER__` can be used with cl.exe too, without harm.

File: Configurations/windows-makefile.tmpl [hunk 7]
```diff
@@ -773,13 +773,24 @@ EOF
...
 EOF
+                } else {
+                    return <<"EOF";
+$target: "$gen0" $deps
+	cmd /C "set "ASM=\$(AS)" & $generator \$@.S"
+    \$(CPP) $incs $cppflags $defs \$@.S > \$@.i
+	move /Y \$@.i \$@
+        del /Q \$@.S
+EOF
+                }
```
Comment (levitte) [inline Configurations/windows-makefile.tmpl:793]: This can also be dropped

File: Configurations/windows-makefile.tmpl (hunk 4 above)
Comment (levitte) [inline Configurations/windows-makefile.tmpl:781]: Drop `/E`, as it's already present (well, `/EP` is...).  Also, don't convert the initial TAB to spaces.

**Review by levitte — COMMENTED** [review]
Another way to look at the whole `__ASSEMBLER__` thing

**Comment by everton1984** [comment]
@levitte Hi! Think I covered all the changes you mentioned, hopefully. Thanks a lot for the review. If you could be so kind to take a look and see what you think while I figure out how to sort the CLA issue.

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
This needs to be rebased on top of fresh master branch.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@paulidale still OK?

**Review by paulidale — APPROVED** [review]

**Comment by hlandau** [comment]
Merged to master and 3.1. Thank you.
